internal/dynamicserver/main.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/alecthomas/kong"
+	"google.golang.org/grpc"
+	pref "google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+var version = "v0.0.0"
+
+type config struct {
+	Address string           `help:"hostname:port" default:"localhost:9090"`
+	Version kong.VersionFlag `short:"V" help:"Print version information" group:"Other:"`
+
+	Service string `help:"Service name" default:"echo.EchoService"`
+	Method  string `help:"Method name" default:"Hello"`
+	In      string `help:"String field of request" default:"message"`
+	Out     string `help:"String field of response proto" default:"response"`
+
+	inMessagDescriptor  pref.MessageDescriptor
+	outMessagDescriptor pref.MessageDescriptor
+}
+
+func main() {
+	cfg := newConfig()
+	ctx := kong.Parse(cfg, kong.Vars{"version": version})
+	fmt.Println("starting server server on", cfg.Address)
+	err := run(cfg)
+	ctx.FatalIfErrorf(err)
+}
+
+func newConfig() *config {
+	in := HelloRequest{}
+	out := HelloResponse{}
+	return &config{
+		inMessagDescriptor:  in.ProtoReflect().Descriptor(),
+		outMessagDescriptor: out.ProtoReflect().Descriptor(),
+	}
+}
+
+func run(cfg *config) error {
+	lis, err := net.Listen("tcp", cfg.Address)
+	if err != nil {
+		return err
+	}
+	s := grpc.NewServer()
+	s.RegisterService(desc(cfg), nil)
+
+	return s.Serve(lis)
+}
+
+func desc(cfg *config) *grpc.ServiceDesc {
+	return &grpc.ServiceDesc{
+		ServiceName: cfg.Service, // "echo.EchoService",
+		HandlerType: nil,
+		Methods: []grpc.MethodDesc{
+			{
+				MethodName: cfg.Method, // "Hello",
+				Handler:    cfg.handler,
+			},
+		},
+	}
+}
+
+func (cfg *config) handler(srv interface{}, ctx context.Context, dec func(interface{}) error, interceptor grpc.UnaryServerInterceptor) (interface{}, error) {
+	if interceptor != nil {
+		return nil, fmt.Errorf("handler with interceptor not implemented")
+	}
+	in := dynamicpb.NewMessage(cfg.inMessagDescriptor)
+	if err := dec(in); err != nil {
+		return nil, err
+	}
+	return cfg.transform(in)
+}
+
+func (cfg *config) transform(in *dynamicpb.Message) (*dynamicpb.Message, error) {
```

Comment:
this function does not return any errors. does it need the error result?

I wonder what ByJSONName does if the json name does not exist as a field?

Reply:
returns nil. just a poc. a dodgy one.

Reply:
this line then panics iirc `val := in.Get(fd)`